### ISSUE TITLE
Optimal Transportation Curve Reconstruction-"CGAL ERROR:assertion violation!"

### ISSUE THREAD

**Issue by hexdforward** [CGAL/cgal#3601]
## Issue Details

An assertion error occurred when using CGAL/Optimal Transportation Curve Reconstruction for 2D point reconstruction. The data of 2D points is very simple. The output of this assertion is as follows: 

```
     terminate called after throwing an instance of 'CGAL::Assertion_exception' 
       what():  CGAL ERROR: assertion violation! 
     Expr: f0 != f1 
     File: /usr/local/include/CGAL/Triangulation_data_structure_2.h 
     Line: 1462 
     Aborted (core dumped) 
```

This issue was discussed recently in the following forum post:
http://cgal-discuss.949826.n4.nabble.com/Optimal-Transportation-Curve-Reconstruction-quot-CGAL-ERROR-assertion-violation-quot-td4664008.html#a4664021

@sloriot tryed shuffling the points and the error disappeared (with and without removing duplicated points) 

The data of 2D point is also given in the end of this issue.
## Source Code

```
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
#include <CGAL/Optimal_transportation_reconstruction_2.h>

#include <vector>
#include <iterator>
#include <iostream>
#include <sstream>
#include <fstream>
#include <unordered_set>

using namespace std;

typedef CGAL::Exact_predicates_exact_constructions_kernel K;
typedef K::FT                                               FT;
typedef K::Point_2                                          Point;
typedef K::Segment_2                                        Segment;
typedef CGAL::Optimal_transportation_reconstruction_2<K>    Otr_2;

void load_xy_file(const std::string& filename, std::vector<Point>& points)
{
  std::ifstream ifs(filename.c_str());
  Point point;
  while (!ifs.eof())
  {
      double a,b,c;
      ifs >> a >> b;
      Point pt(a, b);
      points.push_back(pt);
  }
  ifs.close();
}

void list_output(Otr_2& otr2)
{
  std::cout << "(-------------List output---------- )" << std::endl;

  std::vector<Point> isolated_points;
  std::vector<Segment> segments;

  otr2.list_output(std::back_inserter(isolated_points), std::back_inserter(segments));
  std::vector<Point>::iterator pit;

  std::cout<<"segments size: " << segments.size() << std::endl;
  {
    std::vector<Segment>::iterator sit;
    for (sit = segments.begin(); sit != segments.end(); sit++) 
        cout << (*sit) << endl;
  }
}

int main()
{
    string filepath = "../data/2d_points.txt";
    ifstream infile(filepath);
    std::vector<Point> points;
    load_xy_file(filepath, points);

    Otr_2 otr2(points);
    otr2.set_verbose(1);
    otr2.set_use_flip(1);
    otr2.set_relocation(1);
    otr2.run_under_wasserstein_tolerance(0.8);
    
    list_output(otr2);
    return 0;
}
```
## Environment
* Operating system (): Ubuntu 16.04 
* Compiler: gcc 5.4.0
* CGAL version:  4.13

## Data of 2D Points
```
7.600000 -10.600000
8.400001 -11.400001
6.200000 -8.200000
5.600000 -7.200000
5.800000 -7.200000
5.800000 -7.000000
5.800000 -6.800000
5.600000 -6.400000
5.600000 -6.200000
5.400000 -5.800000
5.400000 -5.600000
5.200000 -5.200000
5.000000 -4.800000
5.000000 -4.600000
5.200000 -4.600000
5.000000 -4.200000
5.000000 -4.000000
5.000000 -3.800000
4.600000 -3.400000
4.800000 -3.400000
4.800000 -3.200000
4.800000 -3.000000
4.400000 -2.600000
4.600000 -2.600000
4.600000 -2.400000
4.600000 -2.200000
4.800000 -2.200000
4.800000 -2.000000
5.000000 -2.000000
5.200000 0.800000
4.600000 0.800000
4.800000 1.000000
4.600000 1.200000
4.600000 1.400000
5.000000 1.600000
5.000000 1.800000
4.800000 1.800000
4.800000 2.000000
4.800000 2.200000
5.000000 2.400000
5.200000 2.600000
5.200000 2.800000
5.400000 3.000000
5.800000 3.400000
12.200000 7.400000
12.400001 7.800000
12.600000 8.000000
13.200000 9.600000
13.000000 9.600000
13.000000 9.800000
13.000000 10.000000
13.200000 10.400001
13.200000 10.600000
13.200000 10.800000
13.200000 11.000000
13.200000 11.200000
13.400001 12.200000
13.600000 12.600000
14.000000 13.200000
12.400001 20.600000
12.200000 20.600000
12.000000 20.400000
12.000000 20.600000
10.200000 29.000000
10.000000 28.800001
9.800000 28.600000
9.800000 28.800001
10.400001 31.600000
10.200000 31.400000
12.600000 43.600002
7.800000 27.400000
7.600000 27.200001
7.400000 27.000000
8.600000 32.600002
8.400001 32.600002
8.200000 32.400002
7.800000 32.400002
7.200000 30.600000
6.200000 26.800001
6.800000 30.000000
6.800000 30.200001
6.600000 30.400000
6.400000 29.600000
6.200000 30.000000
5.800000 29.000000
5.600000 28.400000
5.400000 28.200001
5.200000 28.000000
4.200000 23.800001
4.800000 28.000000
4.600000 27.800001
4.400000 28.000000
4.200000 27.400000
4.000000 27.400000
4.000000 29.000000
4.000000 29.400000
3.400000 28.200001
2.600000 22.800001
5.800000 54.799999
5.600000 54.799999
5.400000 54.400002
5.200000 54.799999
5.000000 54.400002
4.800000 54.200001
4.600000 54.000000
2.200000 27.600000
2.000000 27.400000
1.200000 20.600000
1.000000 20.400000
0.800000 20.200001
0.600000 20.000000
0.400000 20.000000
0.200000 19.800001
0.000000 19.600000
-2.800000 -19.800001
-3.000000 -19.600000
-3.200000 -19.400000
-3.400000 -19.200001
-3.600000 -19.000000
-3.800000 -18.800001
-4.000000 -18.600000
-4.000000 -18.400000
-4.200000 -18.200001
-4.400000 -18.000000
-4.600000 -17.800001
-4.800000 -17.600000
-5.000000 -17.600000
-5.000000 -17.400000
-5.200000 -17.200001
-5.400000 -17.200001
-5.400000 -16.800001
-5.600000 -16.800001
-5.600000 -16.200001
-5.800000 -16.000000
-6.000000 -16.000000
-6.000000 -15.800000
-6.200000 -15.800000
-6.200000 -15.400001
-6.400000 -15.400001
-6.400000 -15.200000
-6.600000 -15.200000
-6.600000 -14.800000
-6.800000 -14.600000
-6.800000 -14.400001
-7.000000 -14.400001
-7.000000 -14.200000
-7.200000 -14.000000
-7.200000 -13.800000
-7.400000 -13.800000
-7.400000 -13.600000
-7.600000 -13.400001
-7.600000 -13.200000
-7.600000 -13.000000
-7.800000 -13.000000
-8.000000 -13.000000
-7.800000 -12.600000
-8.000000 -12.600000
-8.200000 -12.600000
-8.200000 -12.400001
-8.200000 -12.200000
-8.200000 -12.000000
-8.400001 -12.000000
-8.600000 -12.000000
-8.600000 -11.800000
-8.600000 -11.600000
-8.800000 -11.600000
-8.800000 -11.400001
-8.800000 -11.200000
-9.000000 -11.200000
-9.000000 -11.000000
-9.200000 -11.000000
-9.000000 -10.600000
-9.200000 -10.600000
-9.400001 -10.600000
-9.400001 -10.400001
-9.400001 -10.200000
-9.600000 -10.200000
-9.600000 -10.000000
-10.000000 -10.200000
-9.800000 -9.800000
-9.800000 -9.600000
-10.000000 -9.600000
-10.000000 -9.400001
-10.000000 -9.200000
-10.200000 -9.200000
-10.200000 -9.000000
-10.200000 -8.800000
-10.400001 -8.800000
-10.400001 -8.600000
-10.400001 -8.400001
-10.600000 -8.400001
-10.600000 -8.200000
-10.800000 -8.200000
-10.800000 -8.000000
-11.000000 -8.000000
-11.200000 -8.000000
-11.200000 -7.800000
-11.200000 -7.600000
-11.400001 -7.600000
-11.400001 -7.400000
-11.600000 -7.400000
-11.800000 -7.400000
-11.800000 -7.200000
-12.000000 -7.200000
-12.000000 -7.000000
-12.200000 -7.000000
-12.200000 -6.800000
-12.200000 -6.600000
-12.600000 -6.600000
-12.600000 -6.400000
-12.800000 -6.400000
-12.800000 -6.200000
-12.800000 -6.000000
-12.800000 -5.800000
-13.000000 -5.800000
-13.200000 -5.600000
-13.200000 -5.400000
-13.400001 -5.400000
-13.400001 -5.200000
-13.600000 -5.000000
-13.600000 -4.800000
-13.600000 -4.600000
-14.000000 -4.400000
-14.400001 -4.400000
-14.600000 -4.200000
-14.200000 -4.000000
-14.200000 -3.800000
-14.400001 -3.600000
-14.800000 -3.600000
-14.600000 -3.400000
-14.800000 -3.200000
-14.800000 -3.000000
-14.800000 -2.800000
-12.000000 19.400000
-11.600000 19.000000
-11.200000 18.400000
-11.200000 18.600000
-11.000000 18.600000
-11.000000 18.800001
-11.200000 19.400000
-10.800000 19.000000
-11.000000 19.600000
-10.600000 19.200001
-10.800000 19.800001
-10.400001 19.400000
-10.400001 19.600000
-10.200000 19.600000
-9.800000 19.000000
-9.600000 19.000000
-9.400001 18.800001
-9.200000 18.800001
-9.400001 19.400000
-8.800000 18.400000
-8.800000 18.600000
-9.600000 20.400000
-9.200000 20.000000
-8.200000 18.200001
-8.600000 19.200001
-7.800000 18.200001
-7.600000 18.200001
-9.000000 21.800001
-8.800000 22.000000
-8.600000 21.600000
-8.800000 22.400000
-7.400000 19.200001
-7.000000 18.400000
-8.200000 22.000000
-8.200000 22.400000
-7.800000 21.600000
-6.800000 19.200001
-6.600000 19.200001
-6.200000 18.200001
-6.000000 18.000000
-5.800000 18.000000
-5.600000 18.000000
-5.600000 18.200001
-5.400000 18.200001
-5.200000 18.000000
-5.200000 18.200001
-5.000000 18.200001
-4.800000 18.200001
-4.600000 18.200001
-4.400000 18.200001
-4.200000 18.200001
-4.000000 18.200001
-3.800000 18.200001
-3.600000 18.200001
-3.400000 18.200001
-3.200000 18.200001
-3.000000 18.200001
-2.800000 18.400000
-2.600000 18.400000
-2.400000 18.600000
-2.200000 18.600000
-2.000000 18.600000
-1.800000 18.600000
-1.600000 18.800001
-1.400000 18.800001
-1.200000 19.000000
-1.000000 19.400000
-0.800000 19.000000
-0.800000 19.400000
-0.600000 19.400000
-0.400000 19.600000
-0.200000 19.600000
0.000000 19.600000
```

**Comment by johnzjq** [comment]
Randomly shuffling the vertices many times still produce assertion. We found the assertion is triggered during iterations of Join_Vertices and Set_Adjacency. Does it indicate the collapsing operation violates the triangulation at some point? of course, we assumed the initial triangulation is correct.

**Comment by hexdforward** [comment]
Another set of data has produced new assertion errors：

```
reconstruct under tolerance 0.8terminate called after throwing an instance of 'CGAL::Precondition_exception'
  what():  CGAL ERROR: precondition violation!
Expr: b != 0
File: /usr/local/include/CGAL/Lazy_exact_nt.h
Line: 676
Aborted (core dumped)
```

the data of points ：
[4_with_back_80_front_200_.txt](https://github.com/CGAL/cgal/files/2777930/4_with_back_80_front_200_.txt)

**Comment by afabri** [comment]
I didn't look into it in detail, but the `Exact_predicates_exact_constructions_kernel` is probably not appropriate for this package.

**Comment by johnzjq** [comment]
I tried to output the initial triangulation at the first line of decimate() by calling m_dt.file_output(). However, I was confused by the result file. Is it a valid OFF file? Why do the indices of vertices exceed the number of vertices? please see the attached file.
@sgiraudot 
[triangle_out.txt](https://github.com/CGAL/cgal/files/2800256/triangle_out.txt)

**Comment by sgiraudot** [comment]
As I said, the triangulation created by `copy_star()` is invalid, so going from there, there are good chances that many things will be invalid (including incoherence between the real number of vertices and the one returned by `number_of_vertices()`).

If this was not clear: the bug _is_ identified, we just need to find some time to work on it.

**Comment by sgiraudot** [comment]
I had some time to check this out, the bug should be solved by https://github.com/CGAL/cgal/pull/3639.

**Comment by lrineau** [comment]
Fixed by #3639.
